### PR TITLE
Fix openbsd package provider

### PR DESCRIPTION
**Pull request by jdmundrawala**
This provider was broken in just the right ways when  it was merged that it sort of worked. There were references to methods that didn't exist that just happened not to get called ever.

Since we just released this, and it does not work as is, I'm going to change a few things:
- I've simplified the candidate_version search...I didn't quite understand what the original was wanting to do

- This package provider had a weird usage for source (completely inconsistent with the rest of our package providers), and it was not happy with @jaymzh's patch https://github.com/jaymzh/chef/commit/1f380f9f180f39581f2e124061af1ee927a23e0c. I've decided to explicitly make `source` unsupported for this provider

- It now correctly deals with multiple valid candidates by failing the run instead of picking one

Fixes https://github.com/chef/chef/issues/3096

cc @chef/client-core

### CODE REVIEW

File: lib/chef/provider/package/openbsd.rb [hunk 1]
```diff
@@ -84,19 +90,28 @@ def installed_version
...
+              raise Chef::Exceptions::Package, "#{new_resource.name} has multiple matching candidates. You must specify a version" if results.length > 1
             end
-            Chef::Log.debug("candidate_version of '#{@new_resource.package_name}' is '#{result}'")
-            result
           end
         end
 
+        def version_string
+          ver  = ''
+          ver += "-#{new_resource.version}" if new_resource.version
```
Comment (btm) [inline lib/chef/provider/package/openbsd.rb]: @new_resource?

Comment (jdmundrawala) [inline lib/chef/provider/package/openbsd.rb]: https://github.com/chef/chef/blob/master/lib/chef/provider.rb#L55

Comment (jdmundrawala) [inline lib/chef/provider/package/openbsd.rb]: you should point out the one's with @

Comment (btm) [inline lib/chef/provider/package/openbsd.rb]: There are too many of those. Friends jumping off a bridge and all that.

Comment (lamont-granquist) [inline lib/chef/provider/package/openbsd.rb]: accessors are generally better for getters -- `new_resorce` will error out with nomethod, `@new_resorce` will autovivify an ivar.

File: lib/chef/provider/package/openbsd.rb [hunk 2]
```diff
@@ -47,15 +47,21 @@ def load_current_resource
           @current_resource
         end
 
+        def define_resource_requirements
+          super
```
Comment (jaymzh) [inline lib/chef/provider/package/openbsd.rb]: :+1:

File: lib/chef/provider/package/openbsd.rb [hunk 3]
```diff
@@ -84,19 +90,28 @@ def installed_version
...
+            results = results.reject(&:nil?)
+            Chef::Log.debug("candidate versions of '#{@new_resource.package_name}' are '#{results}'")
+            case results.length
+            when 0
+              []
+            when 1
+              results[0]
             else
-              result = pkg_info.stdout[/^#{Regexp.escape(@new_resource.package_name)}-(.+?)\s/, 1]
+              raise Chef::Exceptions::Package, "#{new_resource.name} has multiple matching candidates. You must specify a version" if results.length > 1
```
Comment (jaymzh) [inline lib/chef/provider/package/openbsd.rb]: Every other package provider takes the most-recent if no version is specified (assuming of course the semantics that ':install' does nothing if it's already installed, where as :upgrade will always upgrade to the latest version)

File: lib/chef/provider/package/openbsd.rb (hunk 1 above)
Comment (jaymzh) [inline lib/chef/provider/package/openbsd.rb]: indeed - lamont and I both converted a ton of @new_resource to new_resource's, lets not go back. :)

File: lib/chef/provider/package/openbsd.rb (hunk 3 above)
Comment (jdmundrawala) [inline lib/chef/provider/package/openbsd.rb]: I think version may be the wrong wording here. When querying for something like vim, you get back multiple packages of the same version of vim, just compiled with different flags (x11, python2, python3, etc)

Comment (jaymzh) [inline lib/chef/provider/package/openbsd.rb]: ahhh. That's par tof the name though - vim-x11 or whatever? Perhaps: 'Multiple packages match the name #{name} - please use a more specific name'?

Comment (jdmundrawala) [inline lib/chef/provider/package/openbsd.rb]: now that i think about it, i feel like i've done something wrong. The resource I specify looks like:
```ruby
package 'vim' do
  version '7.4.135p2-no_x11'
end
```
It seems like the `7.4.135p2` should be allowed to float and there should be some way `no-x11` part of it?

Comment (jdmundrawala) [inline lib/chef/provider/package/openbsd.rb]: i see the mistake i've made, i removed some of the the code that did stuff with stems:
http://www.openbsd.org/cgi-bin/man.cgi/OpenBSD-current/man1/pkg_add.1?query=pkg_add&arch=i386

so the right way to define this would be
```ruby
package 'vim--no_x11'
```

File: lib/chef/provider/package/openbsd.rb [hunk 4]
```diff
@@ -84,19 +90,28 @@ def installed_version
 
         def candidate_version
           @candidate_version ||= begin
-            version_string  = ''
-            version_string += "-#{version}" if @new_resource.version
-            pkg_info = shell_out!("pkg_info -I \"#{@new_resource.package_name}#{version_string}\"", :env => nil, :returns => [0,1])
-            if parts = @new_resource.package_name.match(/^(.+?)--(.+)/)
```
Comment (jdmundrawala) [inline lib/chef/provider/package/openbsd.rb]: This needs to go back in

File: lib/chef/provider/package/openbsd.rb (hunk 3 above)
Comment (jdmundrawala) [inline lib/chef/provider/package/openbsd.rb]: I added the flavor stem crap back in

**Comment by jdmundrawala** [comment]
@chef/client-core This kind of works now. I've added the things that do not work to `define_resource_requirements`. These enhancements are available for community contributions, but since openbsd is not a supported platform, I will not be spending more time adding those.

File: lib/chef/provider/package/openbsd.rb (hunk 4 above)
Comment (btm) [inline lib/chef/provider/package/openbsd.rb]: according to @jdmundrawala this went back in a couple lines down.

**Comment by btm** [comment]
:+1: on shipping this with the existing 12.1.2 bugfix releases .

**Comment by lamont-granquist** [comment]
:+1:

**Comment by thommay** [comment]
:+1:

**Comment by danielsdeleo** [comment]
Summary of internal chat discussion: I'm borderline :-1: without some regression tests. There are valid reasons why we can't make the tests as good as we really want to, but a handful of unit tests would go a long way to preventing future regression and documenting the edge cases we've learned we will encounter in the real world.

**Comment by danielsdeleo** [comment]
That said, the code changes seem fine, so throw in some tests and I'm :+1:

File: lib/chef/provider/package/openbsd.rb [hunk 5]
```diff
@@ -71,32 +89,45 @@ def remove_package(name, version)
         private
 
         def installed_version
-          if parts = @new_resource.package_name.match(/^(.+?)--(.+)/)
+          if parts = new_resource.package_name.match(/^(.+?)--(.+)/)
```
Comment (tyler-ball) [inline lib/chef/provider/package/openbsd.rb]: Because this regex is used multiple places, I would like to see it turned into a constant or gotten from a method

**Comment by tyler-ball** [comment]
Yeah, specs then :+1:

**Comment by jdmundrawala** [comment]
I've added more specs.
The missing functionality is marked with pending, so it will fail until somebody implements it

**Comment by danielsdeleo** [comment]
:shipit:

File: lib/chef/provider/package/openbsd.rb [hunk 6]
```diff
@@ -37,25 +38,42 @@ class Openbsd < Chef::Provider::Package
...
+          end
+          requirements.assert(:all_actions) do |a|
+            a.assertion do
+              if new_resource.package_name =~ /^(.+?)--(.+)/
+                !new_resource.version
+              else
+                true
+              end
+            end
+            a.failure_message(Chef::Exceptions::Package, 'The openbsd package provider does not support providing a version and flavor')
```
Comment (jaymzh) [inline lib/chef/provider/package/openbsd.rb]: I'm not sure I understand this. why doesn't this support vim-x11 1.2.4?

Comment (jdmundrawala) [inline lib/chef/provider/package/openbsd.rb]: Because it never worked. The person that wrote the code didn't account for
this case. I'm just making it explicitly clear so you don't get some crazy
exception

**Comment by tyler-ball** [comment]
:+1:

**Comment by btm** [comment]
:+1:
